resources/lib/library.py

```diff
@@ -1,6 +1,7 @@
 # coding=utf-8
 import os, sys, datetime, unicodedata
 import xbmc, xbmcgui, xbmcvfs, urllib
+import xbmcaddon
```

Comment:
Include with xbmc includes on previous line

Reply:
It's funny that you tell me this, the whole project code does NOT meet the current python standards...and in any case putting the imports in line is wrong, you have to study the PEP8 standard

Reply:
> > Lets first ignore...
> > a crash of the addon as it forces the loading of paths before the correct execution of all things
> > ... if an addon is crashing because we've called it too early, then the addon is badly coded. This isn't our issue. For me, that means we can also ignore ...
> 
> there are possible examples of what happens when widgets make the call to the addons when they are executed at the opening of Kodi.
> I have already corrected my addon, in case the user starts the addon before the service, the user is blocked by a message. But of course the widget calls the frontend immediately when opening kodi, always showing users this message, and all the widgets break. This is because it does not respect the services.

This remains not our issue. This script isn't responsible for calling the add-on at run-time (briefly, the script writes a file which becomes a part of the skin - when Kodi displays that skin file, it is responsible for how it calls the widgets. This script has long since done its job.)

> The 1 minute delay you mentioned is not a solution, because it still does not respect the state of the addon services.

Just spitballing ideas.

> > So, and if I'm understanding right, ignoring the above leaves us purely with a situation where the widget can't provide details because it doesn't have them yet,
> 
> I do not know these projects but from what i see there is a clear problem of how they are managed widgets. in script.skinshortcuts and or script.skin.helper.service, they directly modify the skin files to apply the widgets, but i have not found the point where the call to the addons connected to the widgets is made.

I suspect that's because it's made by Kodi itself when it loads the relevant skin file, not by this script or by Skin Helper.

> in my opinion these scripts should intervene dynamically by call, and so perform a passive loading of widgets, which are activated directly by the services of each addons when they are ready.

That would be a call to be made by Team Kodi, or other people contributing code to Kodi itself, as we don't have any control on how Kodi loads widgets.

> or alternatively,
> you can also add a property to each addon like window.setProperty('service', 'running')
> in this case the script that manages the widgets must verify at intervals the values of these properties in each addon and load the contents of the widgets at when the service of addons are 'running'.

That's not something that a third-party script like ours should be dictating to widget providers. Have widget providers agreed to implement something like this?

> Embuary Helper is not used by all skins, they have to agree to use a unique project, there's too much confusion is counterproductive at all.

It wouldn't need to be if it performs a similar function to Skin Helper in this regard - this script was/is responsible for adding properties to the widget path which were provided by Skin Helper.

> as i wrote above this is a Workaround,
> i don't see any developer who's helping,
> are you free to accept or not, this temporary solution
> 
> I'm not the maintainer of these two projects (script.skinshortcuts, script.skin.helper.service)
> I've exposed the problem that has been afflicting users for years

It's up to MikeSilvo whether this is incorporated here. If I'm understanding your issue correctly, the problem with how widgets is called is actually in core Kodi, so you would need to approach Team Kodi to see if any developer is working on fixing the core issue.

Reply:
We're very aware the project doesn't conform to PEP8. A few people have taken on the challenge over the years of updating it to do so, but have never completed it. If you'd like to be the one who overhauls its coding conventions, that would be fantastic :)